core/server/common/src/main/java/alluxio/master/AbstractMaster.java

```diff
@@ -313,4 +323,87 @@ public void close() {
       }
     }
   }
+
+  @NotThreadSafe
+  public final class MasterAuditContext implements AuditContext {
+    private final AsyncUserAccessAuditLogWriter mAsyncAuditLogWriter;
+    private boolean mAllowed;
+    private String mCommand;
+    private String mSrcPath;
+    private String mDstPath;
+    private String mUser;
+    private String mIp;
+    private String mSrcPathOwner;
+    private String mSrcPathGroup;
+    private short mSrcPathMode;
+
+    @Override
+    public MasterAuditContext setAllowed(boolean allowed) {
+      mAllowed = allowed;
+      return this;
+    }
+
+    public MasterAuditContext setCommand(String command) {
+      mCommand = command;
+      return this;
+    }
+
+    public MasterAuditContext setSrcPath(String srcPath) {
+      mSrcPath = srcPath;
+      return this;
+    }
+
+    public MasterAuditContext setDstPath(String dstPath) {
+      mDstPath = dstPath;
+      return this;
+    }
+
+    public MasterAuditContext setUser(String user) {
+      mUser = user;
+      return this;
+    }
+
+    public MasterAuditContext setIp(String ip) {
+      mIp = ip;
+      return this;
+    }
+
+    public MasterAuditContext setSrcPathOwner(String owner) {
+      mSrcPathOwner = owner;
+      return this;
+    }
+
+    public MasterAuditContext setSrcPathGroup(String group) {
+      mSrcPathGroup = group;
+      return this;
+    }
+
+    public MasterAuditContext setSrcPathMode(short mode) {
+      mSrcPathMode = mode;
+      return this;
+    }
+
+    private MasterAuditContext(AsyncUserAccessAuditLogWriter asyncAuditLogWriter) {
+      mAsyncAuditLogWriter = asyncAuditLogWriter;
+      mAllowed = true;
+    }
+
+    @Override
+    public void append() {
+      if (mAsyncAuditLogWriter != null) {
+        mAsyncAuditLogWriter.append(this);
+      }
+    }
+
+    @Override
+    public void close() {}
```

Comment:
I think the current code is not as clean as it could be. take this as example:
```java
  public FileInfo getFileInfo(AlluxioURI path, GetStatusOptions options)
      throws FileDoesNotExistException, InvalidPathException, AccessControlException {
    Metrics.GET_FILE_INFO_OPS.inc();
    try (JournalContext journalContext = createJournalContext();
         AuditContext auditContext = createAuditContext();
         LockedInodePath inodePath = mInodeTree.lockInodePath(path, InodeTree.LockMode.READ)) {
      try {
        mPermissionChecker.checkPermission(Mode.Bits.READ, inodePath);
      } catch (AccessControlException e) {
        appendAuditLogContext("getFileInfo", path, null, inodePath.getInode(),
            auditContext.setAllowed(false));
        throw e;
      }
      if (inodePath.fullPathExists()) {
        // The file already exists, so metadata does not need to be loaded.
        appendAuditLogContext("getFileInfo", path, null, inodePath.getInode(), auditContext);
        return getFileInfoInternal(inodePath);
      }
      // ...
      appendAuditLogContext("getFileInfo", path, null, inodePath.getInode(), auditContext);
      return getFileInfoInternal(inodePath);
    }
  }
```

You put `appendAuditLogContext` in multiple exits which is very vulnerable to code chagne (e.g., adding more returns). However, we should leverage closeable if we put append logic in the `close()` method, it can look like to avoid 

```java
  public FileInfo getFileInfo(AlluxioURI path, GetStatusOptions options)
      throws FileDoesNotExistException, InvalidPathException, AccessControlException {
    Metrics.GET_FILE_INFO_OPS.inc();
    try (JournalContext journalContext = createJournalContext();
         AuditContext auditContext = createAuditContext("checkConsistency", srcPath, null /* dstPath */);
         LockedInodePath inodePath = mInodeTree.lockInodePath(path, InodeTree.LockMode.READ)) {
      try {
        mPermissionChecker.checkPermission(Mode.Bits.READ, inodePath);
      } catch (AccessControlException e) {
        auditContext.setAllowed(false));
        throw e;
      }
      if (inodePath.fullPathExists()) {
        // The file already exists, so metadata does not need to be loaded.
        return getFileInfoInternal(inodePath);
      }
      // ...
      return getFileInfoInternal(inodePath);
    }
  }
```
 or even the following to further minimize the change

```java
  public FileInfo getFileInfo(AlluxioURI path, GetStatusOptions options)
      throws FileDoesNotExistException, InvalidPathException, AccessControlException {
    Metrics.GET_FILE_INFO_OPS.inc();
    try (JournalContext journalContext = createJournalContext();
         AuditContext auditContext = createAuditContext("checkConsistency", srcPath, null /* dstPath */);
         LockedInodePath inodePath = mInodeTree.lockInodePath(path, InodeTree.LockMode.READ)) {
      mPermissionChecker.checkPermission(Mode.Bits.READ, inodePath);
      if (inodePath.fullPathExists()) {
        // The file already exists, so metadata does not need to be loaded.
        return getFileInfoInternal(inodePath);
      }
      // ...
      return getFileInfoInternal(inodePath);
    } catch (AccessControlException e) {
        auditContext.setAllowed(false));
        throw e;
    }
  }
```